rust/kernel/thread.rs

```diff
@@ -0,0 +1,404 @@
+///
+/// We use double boxing in the implementation, because 1) we need to make sure
+/// the thread argument live longer enough for the new thread to use (box #1),
+/// and 2) `dyn FnOnce()` is a fat pointer that doesn't fit in `*mut c_void`
+/// (box #2).
+struct ClosureCallFunctor;
+
+impl ThreadFunctor<Box<Box<dyn FnOnce() -> KernelResult + Send>>> for ClosureCallFunctor {
+    fn func(arg: Box<Box<dyn FnOnce() -> KernelResult + Send>>) -> KernelResult {
+        // Just calls it.
+        arg()
+    }
+}
+
+/// A kernel thread handle.
+pub struct Thread {
+    /// Pointer to the kernel thread.
+    task: *mut bindings::task_struct,
+}
+
+impl Thread {
+    /// Creates a new thread using a C-style function pointer.
+    ///
+    /// No extra dynamic memory allocation for thread creation. Use this when
+    /// closure allocation overhead is unacceptable or there is already a C
+    /// style thread function.
+    ///
+    /// # Safety
+    ///
+    /// This function actually doesn't dereference `arg` or call `f`, so even if
+    /// the users pass incorrect parameters this function won't run into
+    /// trouble. But if the users provide incorrect `arg` or `f` the new
+    /// thread will corrupt memory or do other unsafe behaviors, so
+    /// make it `unsafe`. Otherwise, it is [`Thread::wake_up`] that should be
+    /// made as `unsafe` because it will trigger the call of the unsafe function
+    /// `f`, this is undesirable given that [`Thread::wake_up`] on a
+    /// [`Thread::try_new`] created thread is safe.
+    ///
+    /// The safety requirements of calling this function are:
+    ///
+    /// - Make sure `arg` is a proper pointer that points to a valid memory
+    ///   location when the new thread begins to run.
+    ///
+    /// - Make sure `f` is a proper function pointer and `f` handles `arg`
+    ///   correctly.
+    ///
+    /// # Context
+    ///
+    /// This function might sleep in `kthread_create_on_node` due to the memory
+    /// allocation and waiting for the completion, therefore do not call this
+    /// in atomic contexts (i.e. preemption-off contexts).
+    pub unsafe fn try_new_extern(
+        name: CStr,
+        f: unsafe extern "C" fn(*mut c_types::c_void) -> c_types::c_int,
+        arg: *mut c_types::c_void,
+    ) -> KernelResult<Self> {
+        let task;
+
+        // SAFETY: `kthread_create_on_node` will copy the content of `name`,
+        // so we don't need to make the `name` live longer.
+        task = ptr_to_result(bindings::kthread_create_on_node(
+            Some(f),
+            arg,
+            bindings::NUMA_NO_NODE,
+            cstr!("%s").as_ptr() as _,
+            name.as_ptr(),
+        ))?;
+
+        // Increases the refcount of the task, so that it won't go away if it
+        // `do_exit`.
+        //
+        // SAFETY: `task` is a proper pointer pointing to a newly created
+        // thread.
+        rust_helper_get_task_struct(task);
```

Comment:
Do we really need this? Isn't it the case that `kthread_create_on_node` already returns a referenced task that we can hold on to?

Reply:
`kthread_create_on_node` will return a thread with refcount=1, and if the thread call `do_exit()`, the refcount will be decreased by 1. If we don't `get_task_struct`, we can't guarantee we still hold a valid reference.

Reply:
Are you sure about this? Look at this example: https://elixir.bootlin.com/linux/latest/source/drivers/block/loop.c#L938

I don't see any calls to `get_task_struct` and yet they continue to use the returned pointer, and even call `kthread_stop` eventually, which in fact calls `put_task_struct`, which seems to imply that the refcount already includes a reference for the caller (which would make sense from an API perspective). Or am I missing a `get_task_struct` somewhere in the example?

Reply:
This is because the thread won't exit itself, so it's fine. If you look into `kthread_stop`, you will find even `kthread_stop` itself call `get_task_struct` before `wake_up_process`, this is because the wakee may call `do_exit` after waken up, if that happens, `kthread_stop` will get an invalid task pointer.

Reply:
Yes, you're right. Thanks!

(here's the place where the task pointer is returned: https://elixir.bootlin.com/linux/latest/source/kernel/kthread.c#L278 -- there is no get here.)